Tasks/FtpUpload/ftpuploadtask.ts

```diff
@@ -0,0 +1,243 @@
+        }).fail(failTask);
+    } else {
+        defer.resolve(null);
+    }
+
+    return defer.promise;
+}
+
+
+async function uploadFiles() {
+    tl.debug('connected to ftp host:' + serverEndpointUrl.host);
+
+    tl.debug('files to process: ' + files.length);
+
+    var defer: Q.Deferred<void> = Q.defer<void>();
+
+    cleanRemoteIfRequired().then(() => {
+        var promises = [];
+        var nestedPromises = [];
+        promises.push(ftpHelper.createRemoteDirectory(remotePath).then(() => {
+            directoriesProcessed++;
+            printStatus('remote directory successfully created/verified: ' + remotePath);
+        })); // ensure root remote location exists
+
+        files.forEach((file) => {
+            tl.debug('file: ' + file);
+            var remoteFile: string = preservePaths ?
+                path.join(remotePath, file.substring(rootFolder.length)) :
+                path.join(remotePath, path.basename(file));
+
+            remoteFile = remoteFile.replace(/\\/gi, "/"); // use forward slashes always
+            tl.debug('remoteFile: ' + remoteFile);
+
+            var stats = tl.stats(file);
+            if (stats.isDirectory()) { // create directories if necessary
+                promises.push(ftpHelper.createRemoteDirectory(remoteFile).then(() => {
+                    directoriesProcessed++;
+                    printStatus('remote directory successfully created/verified: ' + remoteFile);
+                }));
+            } else if (stats.isFile()) { // upload files
+                if (overwrite) {
+                    promises.push(ftpHelper.uploadFile(file, remoteFile).then(() => {
+                        filesUploaded++;
+                        printStatus('successfully uploaded: ' + file + ' to: ' + remoteFile);
+                    }));
+                } else {
+                    promises.push(ftpHelper.remoteExists(remoteFile).then((exists: boolean) => {
+                        if (!exists) {
+                            nestedPromises.push(ftpHelper.uploadFile(file, remoteFile).then(() => {
+                                filesUploaded++;
+                                printStatus('successfully uploaded: ' + file + ' to: ' + remoteFile);
+                            }));
+                        } else {
+                            filesSkipped++;
+                            printStatus('skipping file: ' + file + ' remote: ' + remoteFile + ' because it already exists');
+                        }
+                    }));
+                }
+            }
+        });
+
+        Q.all(promises).then(() => {
+            Q.all(nestedPromises).then(() => {
+                tl.setResult(tl.TaskResult.Succeeded, 'FTP upload successful' + getFinalStatusMessage());
+                ftpClient.end();
+                ftpClient.destroy();
+                defer.resolve(null);
+            }).fail(failTask);
+        }).fail(failTask);
+    }).fail(failTask);
+}
+
+ftpClient.on('ready', uploadFiles);
```

Comment:
error handling is still not right and not structured for it.  You're calling async function which must be awaited.

A better pattern which I suggested in the previous review was async function with try/catch and await, await etc...

Something like:
```javascript




async doWork() {
   ftpClient.on('ready', () => {
       try {
           // move these functions to a lib so the task is clean and readable.
           await cleanRemoteIfRequired();
           await ftpHelper.createRemoteDirectory();
           // this will do Q.all() and *must* call reject in the error case
           await uploadfilesInParallel();
           tl.setResult(succeeded, ...);
       }
       catch(err) {
           // any rejects will end up here.
           tf.setResult(failed, ...);
       }
   })

   // handle error event?  how does it do connect?
   ftpClient.on('error', (err) => {
      tl.setResult(failed, ...);
   })
}

doWork()


```